lib/samson/secrets/vault_client.rb

```diff
@@ -23,121 +15,83 @@ class VaultClient < Vault::Client
+        with_retries { vault.logical.read(key) }
       end
 
+      # different servers have different keys so combine all
       def list(path)
-        @vaults.each_value.flat_map { |vault| vault.logical.list(path) }
+        all = @vaults.each_value.flat_map do |vault|
+          with_retries { vault.logical.list(path) }
+        end
+        all.uniq!
+        all
       end
 
-      # make darn sure on deletes and writes that we try a couple of times.
-      # not going to catch any other excpeptions as we want this to blow up
-      # if anything fails
+      # write to all servers that need this key
       def write(key, data)
-        Vault.with_retries(Vault::HTTPConnectionError, attempts: 5) do
-          vault_instances(key).each { |v| v.logical.write(key, data) }
+        clients(key).each do |v|
+          with_retries { v.logical.write(key, data) }
         end
       end
 
+      # delete from all servers that hold this key
       def delete(key)
-        Vault.with_retries(Vault::HTTPConnectionError, attempts: 5) do
-          vault_instances(key).each { |v| v.logical.delete(key) }
+        clients(key).each do |v|
+          with_retries { v.logical.delete(key) }
         end
       end
 
-      def self.available_instances
-        VaultClient.ensure_config_exists
-        JSON.parse(File.read(VaultClient.vault_config_file)).map { |v| v["vault_instance"] }
-      end
-
-      def config_for(instance_name)
-        vault_hosts.detect { |vault| vault["vault_instance"] == instance_name }
-      end
-
-      def self.ensure_config_exists
-        unless File.exist?(VaultClient.vault_config_file)
-          raise "VAULT_CONFIG_FILE or config/vault.json is required for #{ENV["SECRET_STORAGE_BACKEND"]}"
+      def client(deploy_group)
+        unless name = deploy_group.vault_instance.presence
+          raise "deploy group #{deploy_group.permalink} has no vault_instance configured"
         end
-      end
-
-      def self.vault_config_file
-        ENV['VAULT_CONFIG_FILE'] || Rails.root.join("config/vault.json")
+        unless client = @vaults[name]
+          raise "no vault server found with name #{name}"
+        end
+        client
       end
 
       private
 
-      def vault_hosts
-        @vault_hosts ||= JSON.parse(File.read(VaultClient.vault_config_file))
+      def with_retries(&block)
+        Vault.with_retries(Vault::HTTPConnectionError, attempts: 3, &block)
       end
 
-      # get back the vault instance that's required for this pod.  we'll take
-      # the key, and look up the instance or return all of 'em if it's a global key
-      def vault_instances(key)
-        # parse_secret_key doesn't know about the namespace etc, so strip it off
-        deploy_group = SecretStorage.parse_secret_key(key.split('/', 3).last).fetch(:deploy_group_permalink)
-        if deploy_group == 'global'
-          @vaults.map(&:pop)
+      # local server for deploy-group specific key and all for global key
+      def clients(key)
```

Comment:
Having both `client` and `clients` methods that look up different ways is a little confusing.